cardano-testnet/src/Testnet/Util/Cli.hs

```diff
@@ -0,0 +1,186 @@
+  (Comment a) <> (Comment b) = Comment $ a <> "-" <> b
+
+-- Could also use IO (timestamp).
+makeFilePath :: MonadGen m => FilePath -> Comment -> FilePath -> m FilePath
+makeFilePath path comment suffix = do
+  almostUnique <- Comment <$> replicateM 4 hexit
+  return $ path </> unComment ( comment <> almostUnique) <.> suffix
+
+data KeyNames = KeyNames
+  { verificationKeyFile :: FilePath
+  , signingKeyFile :: FilePath
+  }
+
+type KeyGen a = H.Integration (File (a VKey), File (a SKey))
+
+-- cliAddressKeyGen :: FilePath -> Comment -> KeyGen
+cliAddressKeyGen :: TmpDir -> KeyNames -> KeyGen Address
+cliAddressKeyGen = shelleyKeyGen "address" "key-gen"
+
+cliStakeAddressKeyGen :: TmpDir -> KeyNames -> KeyGen StakeAddress
+cliStakeAddressKeyGen = shelleyKeyGen "stake-address" "key-gen"
+
+cliNodeKeyGenVrf :: TmpDir -> KeyNames -> KeyGen Vrf
+cliNodeKeyGenVrf = shelleyKeyGen "node" "key-gen-VRF"
+
+cliNodeKeyGenKes :: TmpDir -> KeyNames -> KeyGen Kes
+cliNodeKeyGenKes = shelleyKeyGen "node" "key-gen-KES"
+
+shelleyKeyGen :: String -> String -> TmpDir -> KeyNames -> KeyGen x
+shelleyKeyGen command subCommand tmpDir keyNames = do
+  let
+    vKeyPath = tmpDir </> verificationKeyFile keyNames
+    sKeyPath = tmpDir </> signingKeyFile keyNames
+  execCli_
+      [ command, subCommand
+      , "--verification-key-file", vKeyPath
+      , "--signing-key-file", sKeyPath
+      ]
+  return (File vKeyPath, File sKeyPath)
+
+cliNodeKeyGen
+  :: TmpDir
+  -> FilePath
+  -> FilePath
+  -> FilePath
+  -> H.Integration (File (Operator VKey), File (Operator SKey), File OperatorCounter)
+cliNodeKeyGen tmpDir vkey skey counter = do
+  let
+    vkPath = tmpDir </> vkey
+    skPath = tmpDir </> skey
+    counterPath = tmpDir </> counter
+  execCli_
+    [ "node", "key-gen"
+    , "--cold-verification-key-file", vkPath
+    , "--cold-signing-key-file", skPath
+    , "--operational-certificate-issue-counter-file", counterPath
+    ]
+  return (File vkPath, File skPath, File counterPath)
+
+data Address key
+data Kes key
+data Operator key
+data StakeAddress key
+data Vrf key
+
+data VKey
+data SKey
+
+data OperatorCounter
+data ByronKey
+data ByronAddress
+data ByronDelegationKey
+data ByronDelegationCert
+
+type TmpDir = FilePath
+type TestnetMagic = Int
+newtype File a = File {unFile :: FilePath}
+  deriving (Show, Eq)
+
+-- This is the back-door for creating File resources in parts of
+-- code that have not been refactored.
+-- When all uses of fakeItH are removed the cleanup is done.
+fakeItH :: FilePath -> H.Integration (File a)
+fakeItH filePath = do
+  return $ File filePath
+  {-
+  exists <- liftIO $ fileExists filePath
+  if exists
+    then return $ File filePath
+    else error (filePath <> " does not exist.")
+-}
+
+fakeIt :: FilePath -> File a
+fakeIt = File
+
+getVKeyPath ::  (File (a VKey), File (a SKey)) -> FilePath
+getVKeyPath (File a, _ ) = a
+
+getSKeyPath ::  (File (a VKey), File (a SKey)) -> FilePath
+getSKeyPath (_, File a) = a
```

Comment:
I think I rather just use `fst` and `snd` and `unFile`.

Reply:
Interesting. I guess I had written the type signature `fst` copied it to a `snd` line
and then edited that line to become the function definition...